src/main/java/com/minecolonies/colony/buildings/BuildingFarmer.java

```diff
@@ -4,89 +4,407 @@
+    }
+
+    /**
+     * Retrieves a random field to work on for the farmer.
+     * @return a field to work on.
+     */
+    public Field getFieldToWorkOn()
+    {
+        Collections.shuffle(farmerFields);
+        for(Field field: farmerFields)
+        {
+            if(field.needsWork())
+            {
+                currentField = field;
+                return field;
+            }
+        }
+        return null;
+    }
+
     @Override
-    public String getSchematicName(){ return FARMER; }
+    public String getSchematicName()
+    {
+        return FARMER;
+    }
 
     @Override
-    public int getMaxBuildingLevel(){ return 3; }
+    public int getMaxBuildingLevel()
+    {
+        return MAX_BUILDING_LEVEL;
+    }
 
     @Override
-    public String getJobName(){ return FARMER; }
+    public String getJobName()
+    {
+        return FARMER;
+    }
+
+    /**
+     * Synchronize field list with colony.
+     * @param world the world the building is in.
+     */
+    public void syncWithColony(World world)
+    {
+        if(!farmerFields.isEmpty())
+        {
+            final ArrayList<Field> tempFields = new ArrayList<>(farmerFields);
+
+            for(final Field field: tempFields)
+            {
+                final ScarecrowTileEntity scarecrow = (ScarecrowTileEntity) world.getTileEntity(field.getID());
+                if(scarecrow == null)
+                {
+                    farmerFields.remove(field);
+                    if (currentField != null && currentField.getID() == field.getID())
+                    {
+                        currentField = null;
+                    }
+                }
+                else
+                {
+                    scarecrow.getInventoryField().setCustomName(LanguageHandler.format("com.minecolonies.gui.scarecrow.user", getWorker().getName()));
+                    field.setInventoryField(scarecrow.getInventoryField());
+                    if(currentField != null && currentField.getID() == field.getID())
+                    {
+                        currentField.setInventoryField(scarecrow.getInventoryField());
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Resets the fields to need work again.
+     */
+    public void resetFields()
+    {
+        for(final Field field: farmerFields)
+        {
+            field.setNeedsWork(true);
+            field.calculateSize(getColony().getWorld(), field.getLocation().down());
+        }
+    }
+
+    @Override
+    public void onDestroyed()
+    {
+        super.onDestroyed();
+        for(final Field field: farmerFields)
+        {
+            Field tempField = getColony().getField(field.getID());
+
+            if(tempField != null)
+            {
+                tempField.setTaken(false);
+                tempField.setOwner("");
+                ScarecrowTileEntity scarecrowTileEntity = (ScarecrowTileEntity) getColony().getWorld().getTileEntity(field.getID());
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'scarecrowTileEntity' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)